subsys/net/lib/http/http_app_server.c

```diff
@@ -0,0 +1,1085 @@
+			   void *user_data)
+{
+	struct http_ctx *ctx = user_data;
+
+	if (!user_data_send) {
+		/* This is the token field in the net_context_send().
+		 * If this is not set, then it is TCP ACK messages
+		 * that are generated by the stack. We just ignore those.
+		 */
+		return;
+	}
+
+	if (ctx->state == HTTP_STATE_OPEN && ctx->cb.send) {
+		ctx->cb.send(ctx, status, user_data_send, ctx->user_data);
+	}
+}
+
+int http_send_error(struct http_ctx *ctx, int code,
+		    u8_t *html_payload, size_t html_len)
+{
+	struct net_pkt *pkt = NULL;
+	const char *msg;
+	int ret;
+
+	switch (code) {
+	case 400:
+		msg = HTTP_STATUS_400_BR;
+		break;
+	}
+
+	ret = http_add_header(ctx, &pkt, msg);
+	if (!ret) {
+		return ret;
+	}
+
+	if (html_payload) {
+		ret = http_add_data(ctx, &pkt, html_payload, html_len);
+		if (!ret) {
+			net_pkt_unref(pkt);
+			return ret;
+		}
+	}
+
+	return http_send_data(ctx, pkt, NULL);
+}
+
+int http_send_msg_raw(struct http_ctx *ctx, struct net_pkt *pkt,
+		      void *user_send_data)
+{
+	int ret;
+
+	NET_DBG("[%p] Sending %zd bytes data", ctx, net_pkt_get_len(pkt));
+
+	ret = net_app_send_pkt(&ctx->app_ctx, pkt, NULL, 0, 0,
+			       user_send_data);
+	if (!ret) {
+		/* We must let the system to send the packet, otherwise TCP
+		 * might timeout before the packet is actually sent. This is
+		 * easily seen if the application calls this functions many
+		 * times in a row.
+		 */
+		k_yield();
+	}
+
+	return ret;
+}
+
+int http_add_header(struct http_ctx *ctx, struct net_pkt **pkt,
+		    const char *http_header)
+{
+	if (!pkt) {
+		return -EINVAL;
+	}
+
+	if (!*pkt) {
+		*pkt = net_app_get_net_pkt(&ctx->app_ctx, AF_UNSPEC,
+					   ctx->timeout);
+		if (!*pkt) {
+			return -ENOMEM;
+		}
+	}
+
+	return _http_add_header(*pkt, ctx->timeout, http_header);
+}
+
+int http_add_data(struct http_ctx *ctx, struct net_pkt **pkt,
+		  const u8_t *buf, size_t len)
+{
+	if (!pkt) {
+		return -EINVAL;
+	}
+
+	if (!*pkt) {
+		*pkt = net_app_get_net_pkt(&ctx->app_ctx, AF_UNSPEC,
+					   ctx->timeout);
+		if (!*pkt) {
+			return -ENOMEM;
+		}
+	}
+
+	return _http_add_chunk(*pkt, ctx->timeout, buf, len);
+}
+
+#if defined(CONFIG_NET_DEBUG_HTTP) && (CONFIG_SYS_LOG_NET_LEVEL > 2)
+static char *sprint_ipaddr(char *buf, int buflen, const struct sockaddr *addr)
+{
+	if (addr->sa_family == AF_INET6) {
+#if defined(CONFIG_NET_IPV6)
+		char ipaddr[NET_IPV6_ADDR_LEN];
+
+		net_addr_ntop(addr->sa_family,
+			      &net_sin6(addr)->sin6_addr,
+			      ipaddr, sizeof(ipaddr));
+		snprintk(buf, buflen, "[%s]:%u", ipaddr,
+			 ntohs(net_sin6(addr)->sin6_port));
+#endif
+	} else if (addr->sa_family == AF_INET) {
+#if defined(CONFIG_NET_IPV4)
+		char ipaddr[NET_IPV4_ADDR_LEN];
+
+		net_addr_ntop(addr->sa_family,
+			      &net_sin(addr)->sin_addr,
+			      ipaddr, sizeof(ipaddr));
+		snprintk(buf, buflen, "%s:%u", ipaddr,
+			 ntohs(net_sin(addr)->sin_port));
+#endif
+	} else {
+		snprintk(buf, buflen, "<invalid address family %d>",
```

Comment:
Although this won't crash, if buflen=NET_IPV4_ADDR_LEN, this will truncate this string if the address family is invalid.

Reply:
Sure, not a big issue thou as this is just used for debug printing.